src/systems/joint_trajectory_controller/JointTrajectoryController.cc

```diff
@@ -0,0 +1,792 @@
+#include <ignition/gazebo/components/ParentEntity.hh>
+#include <ignition/gazebo/components/Name.hh>
+#include <ignition/gazebo/components/Joint.hh>
+#include <ignition/gazebo/components/JointType.hh>
+#include <ignition/gazebo/components/JointPosition.hh>
+#include <ignition/gazebo/components/JointVelocity.hh>
+#include <ignition/gazebo/components/JointForceCmd.hh>
+
+#include <ignition/math/PID.hh>
+
+#include <ignition/transport/Node.hh>
+#include <ignition/transport/Publisher.hh>
+#include <ignition/msgs/joint_trajectory.pb.h>
+#include <ignition/msgs/float.pb.h>
+
+#include "JointTrajectoryController.hh"
+
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+
+/// \brief A single 1-axis joint that is controlled by JointTrajectoryController plugin
+class ActuatedJoint
+{
+  /// \brief Default contructor
+  public: ActuatedJoint() = default;
+
+  /// \brief Constructor that is aware of SDF configuration
+  /// \param[in] _entity Entity of the joint
+  /// \param[in] _sdf SDF reference used to obtain configuration for this joint
+  /// \param[in] _jointIndex Index of the joint, used to determine what SDF parameters belong to it
+  public: ActuatedJoint(const Entity &_entity,
+                        const std::shared_ptr<const sdf::Element> &_sdf,
+                        const size_t &_jointIndex);
+
+  /// \brief Setup components required for control of this joint
+  /// \param[in,out] _ecm Ignition Entity Component Manager
+  /// \return True if setup was successful, False if any of the components could not be created
+  public: bool SetupComponents(ignition::gazebo::EntityComponentManager &_ecm) const;
+
+  /// \brief Set target of the joint that the controller will attempt to reach
+  /// \param[in] _targetPoint Targets of all controlled joint
+  /// \param[in] _jointIndex Index of the joint, used to determine what index of `_targetPoint`
+  /// to use
+  public: void SetTarget(const ignition::msgs::JointTrajectoryPoint &_targetPoint,
+                         const size_t &_jointIndex);
+
+  /// \brief Update command force that is applied on the joint
+  /// \param[in,out] _ecm Ignition Entity Component Manager
+  /// \param[in] _dt Time difference to update for
+  public: void Update(ignition::gazebo::EntityComponentManager &_ecm,
+                      const std::chrono::steady_clock::duration &_dt);
+
+  /// \brief Reset the target of the joint
+  public: void ResetTarget();
+
+  /// \brief Reset the position and velocity PID error on the joint
+  public: void ResetPIDs();
+
+  /// \brief Entity of the joint
+  public: Entity entity;
+
+  /// \brief Target state that the joint controller should reach
+  public: struct TargetState
+  {
+    /// \brief Target position of the joint
+    double position;
+    /// \brief Target position of the joint
+    double velocity;
+    /// \brief Target acceleration of the joint
+    /// \attention Acceleration control is NOT implemented at the moment
```

Comment:
Can we ticket an issue for this?

Also, we may want to print a warning in case the user sends acceleration commands.

Reply:
> Can we ticket an issue for this?

I guess so, but I am not sure in which way that would be implemented. Having a `JointAcceleration` component could be a start?

> Also, we may want to print a warning in case the user sends acceleration commands.

Warning added in https://github.com/ignitionrobotics/ign-gazebo/pull/473/commits/8e3f7a192d13e53c81e9c9f87367d6e6c1b78f65

Reply:
Yeah maybe `JointAccelerationCmd` for the commands. Thanks!